extension-androidauto/build.gradle.kts

```diff
@@ -25,7 +25,7 @@ dependencies {
   compileOnly(project(":sdk"))
   testImplementation(project(":sdk"))
 
-  implementation(Dependencies.googleCarAppLibrary)
+  api(Dependencies.googleCarAppLibrary)
```

Comment:
We have been discussing a `compileOnly` here to make downstream specify their release.

1tap is already using [1.2.0-rc01](https://developer.android.com/jetpack/androidx/releases/car-app). Considering this maps extension is a lower level library that is far from completing the [compatibility checklist](https://github.com/mapbox/mapbox-navigation-android-examples/issues/81) - it needs a navigation, geocoding, and simulation in order to be accepted by the play store.

I think we should make this `compileOnly`.

Reply:
As I mentioned in the PR description:

This PR also makes the GoogleCarApp library as an api dependency, so that users don't have to include the GoogleCarApp dependency themselves, users can still provide their own version of the GoogleCarApp library in their app, in that case, [Gradle dependency resolution](https://docs.gradle.org/current/userguide/dependency_resolution.html#sec:version-conflict) will take place.

> Gradle will consider all requested versions, wherever they appear in the dependency graph. Out of these versions, it will select the highest one.

Reply:
so if user specify Car app library 1.2.0-rc01 together with the android-auto extension 0.1.0, 1.2.0-rc01  will be resolved and used since it's higher version.

Reply:
Here's the dependency graph if I add `implementation("androidx.car.app:app:1.2.0-rc01")`  in the android-auto-app module:

```
+--- project :android-auto-app (*)
+--- junit:junit:4.13
|    \--- org.hamcrest:hamcrest-core:1.3
+--- androidx.car.app:app:1.2.0-rc01
|    +--- androidx.annotation:annotation:1.2.0
|    +--- androidx.core:core:1.7.0
|    |    +--- androidx.annotation:annotation:1.2.0
|    |    +--- androidx.annotation:annotation-experimental:1.1.0
|    |    +--- androidx.lifecycle:lifecycle-runtime:2.3.1
|    |    |    +--- androidx.arch.core:core-runtime:2.1.0
|    |    |    |    +--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    |    |    \--- androidx.arch.core:core-common:2.1.0
|    |    |    |         \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    |    +--- androidx.lifecycle:lifecycle-common:2.3.1
|    |    |    |    \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    |    +--- androidx.arch.core:core-common:2.1.0 (*)
|    |    |    \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    +--- androidx.versionedparcelable:versionedparcelable:1.1.1
|    |    |    +--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    |    \--- androidx.collection:collection:1.0.0 -> 1.1.0
|    |    |         \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    +--- androidx.collection:collection:1.0.0 -> 1.1.0 (*)
|    |    \--- androidx.concurrent:concurrent-futures:1.0.0
|    |         +--- com.google.guava:listenablefuture:1.0
|    |         \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    +--- androidx.lifecycle:lifecycle-viewmodel:2.2.0 -> 2.3.1
|    |    \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    +--- androidx.activity:activity:1.2.0 -> 1.2.3
|    |    +--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    +--- androidx.core:core:1.1.0 -> 1.7.0 (*)
|    |    +--- androidx.lifecycle:lifecycle-runtime:2.3.1 (*)
|    |    +--- androidx.lifecycle:lifecycle-viewmodel:2.3.1 (*)
|    |    +--- androidx.savedstate:savedstate:1.1.0
|    |    |    +--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    |    +--- androidx.arch.core:core-common:2.0.1 -> 2.1.0 (*)
|    |    |    \--- androidx.lifecycle:lifecycle-common:2.0.0 -> 2.3.1 (*)
|    |    +--- androidx.lifecycle:lifecycle-viewmodel-savedstate:2.3.1
|    |    |    +--- androidx.annotation:annotation:1.0.0 -> 1.2.0
|    |    |    +--- androidx.savedstate:savedstate:1.1.0 (*)
|    |    |    +--- androidx.lifecycle:lifecycle-livedata-core:2.3.1
|    |    |    |    +--- androidx.arch.core:core-common:2.1.0 (*)
|    |    |    |    +--- androidx.arch.core:core-runtime:2.1.0 (*)
|    |    |    |    \--- androidx.lifecycle:lifecycle-common:2.3.1 (*)
|    |    |    \--- androidx.lifecycle:lifecycle-viewmodel:2.3.1 (*)
|    |    +--- androidx.collection:collection:1.0.0 -> 1.1.0 (*)
|    |    \--- androidx.tracing:tracing:1.0.0
|    |         \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    +--- androidx.lifecycle:lifecycle-common-java8:2.2.0
|    |    +--- androidx.lifecycle:lifecycle-common:2.2.0 -> 2.3.1 (*)
|    |    \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    +--- androidx.annotation:annotation-experimental:1.1.0
|    \--- org.jetbrains.kotlin:kotlin-stdlib:1.6.10
|         +--- org.jetbrains:annotations:13.0
|         \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.6.10
+--- org.jetbrains.kotlin:kotlin-android-extensions-runtime:1.5.31
|    \--- org.jetbrains.kotlin:kotlin-stdlib:1.5.31 -> 1.6.10 (*)
+--- project :extension-androidauto
|    +--- androidx.car.app:app:1.1.0 -> 1.2.0-rc01 (*)
|    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.5.31 -> 1.6.0
|    |    \--- org.jetbrains.kotlin:kotlin-stdlib:1.6.0 -> 1.6.10 (*)
|    +--- androidx.core:core-ktx:1.3.1
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.3.71 -> 1.6.10 (*)
|    |    +--- androidx.annotation:annotation:1.1.0 -> 1.2.0
|    |    \--- androidx.core:core:1.3.1 -> 1.7.0 (*)
|    \--- androidx.annotation:annotation:1.1.0 -> 1.2.0
```

As you see 
```
+--- project :extension-androidauto
|    +--- androidx.car.app:app:1.1.0 -> 1.2.0-rc01 (*)
```
The `androidx.car.app` is resolved as `1.2.0-rc01`

Reply:
> [Resolution strategy](https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:resolution-strategy) [Apache Maven](https://maven.apache.org/) uses a nearest first strategy.

Nice link on the dependency resolution. Sorry I didn't read the full description before commenting 🤦 .

Looks like it specifically chooses the "nearest" not the "highest"

```
myapp > mapbox:maps:extension:0.1.0 > car:app:1.1.0
myapp > car:app:1.0.0: > mapbox:maps:extension:0.1.0 > car:app:1.1.0
```
For example this could downgrade to version `1.0.0` a downstream dependency could downgrade the version if it wants. 

It seems we have the same rules `mapbox:maps`. If downstream wants to upgrade or downgrade the map version they are free to do so if we include the map sdk with `api`

Reply:
Sticking to a minimum version of maps car so they can always update to highest for car library. 👍